docs/builtin-objects.md

```diff
@@ -0,0 +1,56 @@
+# helmfile template built-in objects
+
+- `Environment`: The information about the environment. This is set by the
+  `--environment` flag. It has several objects inside of it:
+  - `Environment.Name`: The name of the environment
+- `StateFile`: The information about the state file. It has several objects
+  inside of it:
+  - `StateFile.Name`: The name of the current state file
+  - `StateFile.BasePath`: The base path of the current state file
+- `RootStateFile`: The information about the root state file. It has several objects
+  inside of it:
+  - `StateFile.Path`: The path of the root state file
```

Comment:
@yxxhero It might be nicer if we could remove this RootStateFile for now, because it might introduce the ability for a sub-helmfile to depend on a dynamically determined "upper-level" helmfile, which can make our philosophy of ensuring each sub-helmfile independently runnable. WDYT?

Reply:
introduce the ability for a sub-helmfile to depend on a dynamically determined "upper-level" helmfile. this a the target of the PR. 😃  but if you disagree with this. I will update the code.

Reply:
@mumoshu

Reply:
And each sub-helmfile still is independent after the PR. 😄

Reply:
Thanks for confirming @yxxhero!

> And each sub-helmfile still is independent after the PR

It is not if we merge this PR as-is, because now it's possible to make a helmfile work only when there's a specific root helmfile exist. For example, you refer to commands used in "exec" under a path relative to the root helmfile, it might work differently depending on where the actual root helmfile is located.

Reply:
yeah. make sense. I will update the code.

Reply:
I was somehow confused if this is something for #623.
Assuming this is actually for #333, please check my response at https://github.com/helmfile/helmfile/discussions/333#discussioncomment-4813437